jena-core/src/main/java/com/hp/hpl/jena/reasoner/rulesys/impl/LPBRuleEngine.java

```diff
@@ -252,19 +272,31 @@ public synchronized void tablePredicate(Node predicate) {
     /**
      * Return a generator for the given goal (assumes that the caller knows that
      * the goal should be tabled).
+     * 
+     * Note: If an earlier Generator for the same <code>goal</code> exists in the
+     * cache, it will be returned without considering the provided <code>clauses</code>.
+     * 
      * @param goal the goal whose results are to be generated
      * @param clauses the precomputed set of code blocks used to implement the goal
      */
-    public synchronized Generator generatorFor(TriplePattern goal, List<RuleClauseCode> clauses) {
-        Generator generator = tabledGoals.get(goal);
-        if (generator == null) {
-            LPInterpreter interpreter = new LPInterpreter(this, goal, clauses, false);
-            activeInterpreters.add(interpreter);
-            generator = new Generator(interpreter, goal);
-            schedule(generator);
-            tabledGoals.put(goal, generator);
-        }
-        return generator;
+    public synchronized Generator generatorFor(final TriplePattern goal, final List<RuleClauseCode> clauses) {
+		return getCachedTabledGoal(goal, new Callable<Generator>() {
+		 	@Override
+		    public Generator call() {
+		 		/** FIXME: Unify with #generatorFor(TriplePattern) - but investigate what about
+		 		 * the edge case that this method might have been called with the of goal == null
+		 		 * or goal.size()==0 -- which gives different behaviour in 
+		 		 * LPInterpreter constructor than through the route of
+		 		 * generatorFor(TriplePattern) which calls a different LPInterpreter constructor
+		 		 * which would fill in from RuleStore. 
+		 		 */  
```

Comment:
I don't know enough about the rules engine to comment.  We need another reviewer here.